File: infra/gcp_auth.py
"""
GCP Service Account authentication helper.
Returns an OAuth2 access token for use with GCP REST APIs.
"""

import json
import time
import jwt
import requests

KEY_FILE = "/home/user/cosmic-backend/infra/gcp-service-account-key.json"

def get_access_token(scopes=None):
    """Get OAuth2 access token from service account key."""
    if scopes is None:
        scopes = "https://www.googleapis.com/auth/cloud-platform"

    with open(KEY_FILE) as f:
        sa = json.load(f)

    now = int(time.time())
    payload = {
        "iss": sa["client_email"],
        "sub": sa["client_email"],
        "aud": sa["token_uri"],
        "iat": now,
        "exp": now + 3600,
        "scope": scopes,
    }

    signed_jwt = jwt.encode(payload, sa["private_key"], algorithm="RS256")

    resp = requests.post(sa["token_uri"], data={
        "grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer",
        "assertion": signed_jwt,
    })
    resp.raise_for_status()
    return resp.json()["access_token"]


if __name__ == "__main__":
    token = get_access_token()
    print(token)

File: infra/gcp_auth.mjs
/**
 * GCP Service Account authentication via JWT.
 * Returns an OAuth2 access token for GCP REST APIs.
 *
 * Usage:
 *   node infra/gcp_auth.mjs                    # prints token
 *   node infra/gcp_auth.mjs --curl URL          # curl with auth header
 */

import { readFileSync } from "fs";
import { createSign } from "crypto";

const KEY_FILE = "/home/user/cosmic-backend/infra/gcp-service-account-key.json";

export async function getAccessToken() {
  const sa = JSON.parse(readFileSync(KEY_FILE, "utf8"));

  const now = Math.floor(Date.now() / 1000);
  const header = { alg: "RS256", typ: "JWT" };
  const payload = {
    iss: sa.client_email,
    sub: sa.client_email,
    aud: sa.token_uri,
    iat: now,
    exp: now + 3600,
    scope: "https://www.googleapis.com/auth/cloud-platform",
  };

  const b64 = (obj) =>
    Buffer.from(JSON.stringify(obj)).toString("base64url");

  const unsigned = `${b64(header)}.${b64(payload)}`;
  const sign = createSign("RSA-SHA256");
  sign.update(unsigned);
  const signature = sign.sign(sa.private_key, "base64url");
  const jwt = `${unsigned}.${signature}`;

  const resp = await fetch(sa.token_uri, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: `grant_type=urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Ajwt-bearer&assertion=${jwt}`,
  });

  if (!resp.ok) {
    const err = await resp.text();
    throw new Error(`Token exchange failed: ${resp.status} ${err}`);
  }

  const data = await resp.json();
  return data.access_token;
}

// CLI usage
if (process.argv[1]?.endsWith("gcp_auth.mjs")) {
  const token = await getAccessToken();
  if (process.argv[2] === "--curl") {
    // Execute curl with auth
    const url = process.argv[3];
    console.log(`Authorization: Bearer ${token}`);
  } else {
    console.log(token);
  }
}
